Update `sig_stfl.h` document for #1650

ducnguyen-sb opened the pull request:
This PR address the comments in #1650 

This PR rewrite the documentation in `sig_stfl.h` header for a few APIs, make simpler and (somewhat) easier to read. 

<!-- Please give a brief explanation of the purpose of this pull request. -->

<!-- Does this PR resolve any issue?  If so, please reference it using automatic-closing keywords like "Fixes #123." -->

<!-- Any PR adding a new feature is expected to contain a test; the test should be part of CI testing, preferably within the ".github/workflows" directory tree. Please add an explanation to the PR if/when (why) this cannot be done. -->

<!-- Please answer the following questions to help manage version and changes across projects. -->

* [ ] Does this PR change the input/output behaviour of a cryptographic algorithm (i.e., does it change known answer test values)?  (If so, a version bump will be required from *x.y.z* to *x.(y+1).0*.)
* [ ] Does this PR change the list of algorithms available -- either adding, removing, or renaming? Does this PR otherwise change an API? (If so, PRs in fully supported downstream projects dependent on these, i.e., [oqs-provider](https://github.com/open-quantum-safe/oqs-provider) and [OQS-OpenSSH](https://github.com/open-quantum-safe/openssh) will also need to be ready for review and merge by the time this is merged.)

<!-- Once your pull request is ready for review and passing continuous integration tests, please convert from a draft PR to a normal PR, and request a review from one of the OQS core team members. -->

baentsch commented on src/sig_stfl/sig_stfl.h:
```diff
@@ -284,37 +283,46 @@ typedef struct OQS_SIG_STFL_SECRET_KEY {
+	 * @param[in] context Pointer to application-specific data, handled externally, associated with the key.
+	 * @returns OQS_SUCCESS if the deserialization succeeds, with the `sk` object populated with the key material.
+	 *
+	 * @remark The caller must ensure that resources (e.g., memory for `sk_buf`) are properly managed.
```
What is the meaning of "properly managed"? Would it be better to state that explicitly (who/when alloc/free need to be called; is both heap and stack allocation OK?)?

ducnguyen-sb replied:
I updated the doc, but I'm not sure it's right. 
If the `sk_buf` is stack, then user don't have to deallocate. So I say a generic one "properly managed". 
I changed to 
`* @attention The caller is responsible for ensuring that `sk_buf` is securely deallocated when it's no longer needed.` in https://github.com/open-quantum-safe/liboqs/pull/1655/commits/c577c2ed8d6597c1766f5d9612397ed876061aa3

baentsch commented on src/sig_stfl/sig_stfl.h:
```diff
@@ -433,7 +447,7 @@ OQS_API OQS_STATUS OQS_SIG_STFL_sign(const OQS_SIG_STFL *sig, uint8_t *signature
 
 /**
- * Query number of remaining signatures
+ * Query the number of remaining signatures.
```
"remain" undocumented?

ducnguyen-sb replied:
Added the description why the remaining signature and total signature matters in https://github.com/open-quantum-safe/liboqs/commit/c577c2ed8d6597c1766f5d9612397ed876061aa3

baentsch commented on src/sig_stfl/sig_stfl.h:
```diff
@@ -469,105 +483,118 @@ OQS_API void OQS_SIG_STFL_free(OQS_SIG_STFL *sig);
+ * Attach an unlock mechanism to a secret key object.
  *
- * Sets function to prevent multiple processes from using the sk at the same time.
+ * This allows for proper synchronization * in a multi-threaded or multi-process environment,
```
superfluous "*"?

ducnguyen-sb replied:
My bad, removed in https://github.com/open-quantum-safe/liboqs/commit/c577c2ed8d6597c1766f5d9612397ed876061aa3

baentsch commented on src/sig_stfl/sig_stfl.h:
```diff
@@ -469,105 +483,118 @@ OQS_API void OQS_SIG_STFL_free(OQS_SIG_STFL *sig);
+ * Assign a mutex function to handle concurrency control over the secret key.
  *
- * Sets function to prevent multiple processes from using the sk at the same time.
+ * This is to ensure that only one process can access or modify the key at any given time.
```
What is the implication of _not_ setting such mutex? Will the library not function, even in a single-threaded environment?

ducnguyen-sb replied:
Oh right, thanks for the question, I add code to catch this behavior. 
Fixed in 
https://github.com/open-quantum-safe/liboqs/pull/1655/commits/efccbe58a339266838a0617df8562886edc068da

ducnguyen-sb replied:
Updated the doc in https://github.com/open-quantum-safe/liboqs/pull/1655/commits/c577c2ed8d6597c1766f5d9612397ed876061aa3

baentsch commented on src/sig_stfl/sig_stfl.h:
```diff
@@ -469,105 +483,118 @@ OQS_API void OQS_SIG_STFL_free(OQS_SIG_STFL *sig);
- * OQS_SIG_STFL_SECRET_KEY_lock .
- *
- * Locks the secret key so only one application that holds the lock can access it.
+ * Lock the secret key to ensure exclusive access in a concurrent environment.
```
I'd assume this function will fail if no mutex has been set?! Worthwhile stating? When does this function need to be called?

ducnguyen-sb replied:
Caught the case when no mutex is set in https://github.com/open-quantum-safe/liboqs/pull/1655/commits/efccbe58a339266838a0617df8562886edc068da.  
Updated in https://github.com/open-quantum-safe/liboqs/pull/1655/commits/c577c2ed8d6597c1766f5d9612397ed876061aa3

baentsch commented on src/sig_stfl/sig_stfl.h:
```diff
@@ -469,105 +483,118 @@ OQS_API void OQS_SIG_STFL_free(OQS_SIG_STFL *sig);
 
 /**
- * OQS_SIG_STFL_SECRET_KEY_unlock .
+ * Unlock the secret key, making it accessible to other processes.
```
Same question as above: When does (an application) need to call this function?

ducnguyen-sb replied:
So I added: 
```
* This operation is essential in multi-threaded or multi-process contexts
 * in order to prevent simultaneous operations that could compromise the stateful signature security.
 
 ...
* @note It's not required to use lock and unlock functions in a single-threaded environment.

```
So, it's should be call in multi-threaded environment. 
I wonder if the current documentation is good enough, let me know if it's unclear.

baentsch replied:
> I wonder if the current documentation is good enough, let me know if it's unclear.

I can only judge from my perspective and cannot know what other's may (mis)understand, so here goes my question now: Which (`liboqs` API) calls ("operations" in your wording above) need to be guarded with this function in multi-threaded environments? And a much more basic question: If there are such library functions failing to properly operate in a multi-threaded environment, wouldn't it be better to document _them_ accordingly? And as follow-on question: Does it need such `liboqs`-based "lock/unlock" functions at all if such documentation exists? Why isn't use of any external, e.g. as provided by the application-chosen threading library, lock/unlock mechanisms sufficient?

I guess what I'd like to suggest in a more "brutal question": Should we not strive to implement a thread-safe library API OR completely get out of the way (giving the app developer full control of what they want/need to do)? Anything in between IMO has the potential for a "support nightmare" ("use API XYZ if you are using threading library ABC; if not using any threading, disregard XYZ but use API XZZ", etc....)

ducnguyen-sb replied:
I answer each question below. 

1. What function need to be guarded with this function in a concurrent environment? 

Only the Signing operation.

Thanks to your question, I updated this note: 

```
* @note It's not necessary to use this function in either Keygen or Verifying operations.
 *       In a concurrent environment, the user is responsible for locking and unlocking the private key,
 *       to make sure that only one thread can access the private key during a Signing operation.
 ```

2. Why such lock/unlock APIs are there? Is it meant we strive to be a thread-safe library?

These APIs intention is to limit to one owner of private key at a time in a concurrent environment.
If users supply lock/unlock logic with a mutex, we acquire the lock to be the only owner of the private key, and unlock after the Signing operation complete. 
If users don't  supply the lock/unlock logic or a mutex, they are simply NOP operations. 

I think the key different here is: We don't execute stateful signature in multi-thread  at the library itself, but we provide a way to let users use multi-thread securely, in the right way. 

3. Does it need such liboqs-based "lock/unlock" functions at all if such documentation exists? 

I think to have `lock/unlock` functions help liboqs in:

- Testing. Someone has to foreseen the usage of our proposed APIs in multi-thread environment, so we need `lock/unlock` to test if it's still safe in such if our low-level implementation change. 

So we have it in test, well, why not exposed them to library user, it's convenience anyway. So there are `liboqs`-based `lock/unlock` functions.

4. Why isn't use of any external, e.g. as provided by the application-chosen threading library, lock/unlock mechanisms sufficient?

Actually we don't concretely define the `lock/unlock` mechanisms in `liboqs`, if the function pointers of `lock/unlock` are non-NULL then we will run them to perform `lock/unlock` the private key. 

So, we open to any application-chosen threading library, and it's user responsibility to create a proper `lock/unlock` functions, and it's up to them to define their `mutex`. 

5. Anything in between IMO has the potential for a "support nightmare" ("use API XYZ if you are using threading library ABC; if not using any threading, disregard XYZ but use API XZZ", etc....)

I totally agree with you, we don't want to get into "support nightmare". I want to run away from it as far as possible. :)) 

You have more experience than me in this matter, please let me know if there is something I need to change to get out of "support nightmare".

ducnguyen-sb replied:
@baentsch  before I merge this PR, do I answer your questions?

baentsch replied:
> You have more experience than me in this matter, please let me know if there is something I need to change to get out of "support nightmare".

I'm not sure I do -- I did not use these APIs -- yet. My "general" question regarding minimizing support effort is: Could we completely do away with those functions in the external OQS API? If we could, we won't have to support them.

> So we have it in test, well, why not exposed them to library user, it's convenience anyway. So there are liboqs-based lock/unlock functions.

That's what I'm concerned with: If it's mere "convenience" (for us) we shouldn't externally make it available. Any API externally visible creates more support effort. Please see the discussion in #1648 and decide whether mutex/lock/unlock is another candidate for this internal API.

ducnguyen-sb replied:
I double check, these APIs just for testing purpose, so I moved them to internal APIs. 
Updated in https://github.com/open-quantum-safe/liboqs/pull/1655/commits/8b0e15e5ca0772c958a4ac368a8cf0a13d63b70a

ducnguyen-sb replied:
The `set_lock, set_mutex, and set_unlock` should still be public APIs. 
@baentsch

baentsch replied:
:-( I won't respond to user questions about them.

baentsch reviewed (commented):
Thanks for the updates addressing many of my original comments, @ducnguyen-sb ! I only re-reviewed the changes in the PR, not the merge-to-main PR again: Please let me know when I should do that (to close out/mark as resolved my comments there).

ducnguyen-sb commented:
Thanks @baentsch , I addressed your comments. 
I believe it's simpler to address a smaller changes here rather than the big PR in #1650 , once it's resolved here we can mark the comments in the main branch as resolved.
